### PR TITLE
Add doxygen

### PR DESCRIPTION
**Pull request by yasahi-hpc**
In this PR, I have added a template for documentation with doxygen + sphinx. Workflow may be added based on this PR.

### CODE REVIEW

File: docs/Makefile
Comment (jbigot) [inline docs/Makefile:1]: Why is a Makefile needed in addition to cmake?

File: docs/conf.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+# If extensions (or modules to document with autodoc) are in another directory,
+# add these directories to sys.path here. If the directory is relative to the
+# documentation root, use os.path.abspath to make it absolute, like shown here.
+#
+# import os
+# import sys
+# sys.path.insert(0, os.path.abspath('.'))
+import subprocess, os
+
+def configureDoxyfile(input_dir, output_dir, doxyfile_in, doxyfile_out):
```
Comment (jbigot) [inline docs/conf.py:18]: this same functionality is present in CMake, why is there a duplication?

File: docs/conf.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    ###                        -DKokkosFFT_ENABLE_DOCS=ON ..'
+    ###build = 'cmake --build . -j 2'
+    ###                        
+    ###subprocess.call(f'cd ../; mkdir build; cd build; {cmake_commands}; {build}', shell=True)
+    input_dir = cwd + '/..'
+    output_dir = cwd +'/doxygen/'
+    doxyfile_in = cwd + '/Doxyfile.in'
+    doxyfile_out = cwd + '/Doxyfile'
+    configureDoxyfile(input_dir, output_dir, doxyfile_in, doxyfile_out)
+    subprocess.call('pwd; ls -lat; doxygen Doxyfile; ls -lat doxygen/xml', shell=True)
```
Comment (jbigot) [inline docs/conf.py:63]: here again I don't understand, does CMake handle Doxygen call or python?

File: fft/src/KokkosFFT_Plans.hpp [hunk 3]
```diff
@@ -39,34 +45,72 @@ namespace Impl {
...
+
+  //! directions of fft
   KokkosFFT::Impl::Direction m_direction;
 
-  // Keep extents
+  //! extents of input/output views
   extents_type m_in_extents, m_out_extents;
 
-  // Only used when transpose needed
+  //! Internal buffers used for transpose
```
Comment (jbigot) [inline fft/src/KokkosFFT_Plans.hpp:101]: without `@{` `@}`, you only comment the first next variable `m_in_T`, not the following one `m_out_T`

**Review by jbigot — APPROVED** [review]
Great documentation effort! This looks great to me.

File: docs/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+set(DOXYFILE_OUT ${CMAKE_CURRENT_BINARY_DIR}/Doxyfile)
+
+# Set the Doxygen input and output directories in the Doxyfile
+configure_file(${DOXYFILE_IN} ${DOXYFILE_OUT})
+
+# Doxygen won't create this for us
+file(MAKE_DIRECTORY ${DOXYGEN_OUTPUT_DIR}) 
+
+# Only regenerate Doxygen when the Doxyfile or public headers change
+add_custom_command(OUTPUT ${DOXYGEN_INDEX_FILE}
```
[the remark this answers is not in the document]
Comment (yasahi-hpc) [reply docs/CMakeLists.txt:23]: I did not notice this command. Seems simpler.

File: docs/Makefile
Comment (yasahi-hpc) [reply docs/Makefile:1]: It is for sphinx, and in the end, we do not rely on CMake in the rtd workflow.

File: docs/conf.py (hunk 2 above)
Comment (yasahi-hpc) [reply docs/conf.py:65]: In the rtd workflow, only the python works. CMake is not used.

File: fft/src/KokkosFFT_Plans.hpp (hunk 3 above)
Comment (yasahi-hpc) [reply fft/src/KokkosFFT_Plans.hpp:101]: So, we need the comment like this?

`//@{ //! Internal buffers used for transpose @}//`

Comment (jbigot) [reply fft/src/KokkosFFT_Plans.hpp:101]: not exactly, the `@{` and `@}` should surround the variables, see https://www.doxygen.nl/manual/grouping.html

Comment (yasahi-hpc) [reply fft/src/KokkosFFT_Plans.hpp:101]: So, 

```
///@{ 
  //! extents of input/output views
  extents_type m_in_extents, m_out_extents;
///@}
```

File: fft/src/KokkosFFT_Plans.hpp [hunk 5]
```diff
@@ -39,34 +45,72 @@ namespace Impl {
...
   nonConstInViewType m_in_T;
   nonConstOutViewType m_out_T;
```
Comment (jbigot) [inline fft/src/KokkosFFT_Plans.hpp:103]:
```suggestion
  //! @{
  //! Internal buffers used for transpose
  nonConstInViewType m_in_T;
  nonConstOutViewType m_out_T;
  //! @}
```

**Pull request merged**